### PR TITLE
[Core] Use numpy to speed up padded token processing

### PR DESCRIPTION
**Pull request by peng1999**
This small change replaces list comprehension with NumPy operations. It results in approximately a 50% speedup for the sampler when processing large batch sizes. This improvement will enhance throughput when running small models. (See benchmark results below)

Related: #249 #5289

## Benchmarks

The following test was conducted on an NVIDIA A100 80GB GPU.

### Sampler latency

This specifically measures the latency of the sampler component.

图片

When prompt token is long, the optimization will have large benefits.

图片

For short prompts, the optimization yields a modest speedup. However, it still outperforms the original version.

### Throughput

The output is obtained by running the following script:

```py
from vllm import LLM, SamplingParams

model_path = "/mnt/data/models/Qwen2-0.5B-Instruct"
llm = LLM(model=model_path, disable_log_stats=False, max_num_seqs=2048, trust_remote_code=True)
tokenizer = llm.get_tokenizer()

with open("text.txt", "r") as file:
    text = file.read()
length = 1024
text = tokenizer.decode(tokenizer.encode(text)[:length])
prompts = [text] * 2048
sampling_params = SamplingParams(temperature=1.0, top_p=0.5, top_k=20, max_tokens=128)

output = llm.generate(prompts, sampling_params)
```

<details>
<summary>main version</summary>

```
Processed prompts:   0%|                                                                                   | 0/2048 [00:00<?, ?it/s]INFO 07-15 17:06:37 metrics.py:334] Avg prompt throughput: 5243.8 tokens/s, Avg generation throughput: 5.1 tokens/s, Running: 32 reqs, Swapped: 0 reqs, Pending: 2016 reqs, GPU KV cache usage: 0.6%, CPU KV cache usage: 0.0%
INFO 07-15 17:06:42 metrics.py:334] Avg prompt throughput: 192507.0 tokens/s, Avg generation throughput: 188.0 tokens/s, Running: 992 reqs, Swapped: 0 reqs, Pending: 1056 reqs, GPU KV cache usage: 19.1%, CPU KV cache usage: 0.0%
INFO 07-15 17:06:47 metrics.py:334] Avg prompt throughput: 187176.6 tokens/s, Avg generation throughput: 182.8 tokens/s, Running: 1920 reqs, Swapped: 0 reqs, Pending: 128 reqs, GPU KV cache usage: 36.9%, CPU KV cache usage: 0.0%
INFO 07-15 17:06:52 metrics.py:334] Avg prompt throughput: 26000.9 tokens/s, Avg generation throughput: 3681.8 tokens/s, Running: 2048 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 40.0%, CPU KV cache usage: 0.0%
INFO 07-15 17:06:57 metrics.py:334] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 3821.7 tokens/s, Running: 2045 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 40.6%, CPU KV cache usage: 0.0%
INFO 07-15 17:07:02 metrics.py:334] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 3997.1 tokens/s, Running: 2045 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 40.6%, CPU KV cache usage: 0.0%
INFO 07-15 17:07:08 metrics.py:334] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 3947.8 tokens/s, Running: 2043 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 41.2%, CPU KV cache usage: 0.0%
INFO 07-15 17:07:13 metrics.py:334] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 3580.9 tokens/s, Running: 2043 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 41.8%, CPU KV cache usage: 0.0%
INFO 07-15 17:07:19 metrics.py:334] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 3820.1 tokens/s, Running: 2043 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 41.8%, CPU KV cache usage: 0.0%
INFO 07-15 17:07:24 metrics.py:334] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 3637.1 tokens/s, Running: 2043 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 42.4%, CPU KV cache usage: 0.0%
INFO 07-15 17:07:29 metrics.py:334] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 3586.9 tokens/s, Running: 2042 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 42.4%, CPU KV cache usage: 0.0%
INFO 07-15 17:07:34 metrics.py:334] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 3468.2 tokens/s, Running: 2041 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 43.0%, CPU KV cache usage: 0.0%
INFO 07-15 17:07:39 metrics.py:334] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 3477.8 tokens/s, Running: 2041 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 43.0%, CPU KV cache usage: 0.0%
INFO 07-15 17:07:45 metrics.py:334] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 3415.1 tokens/s, Running: 2041 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 43.6%, CPU KV cache usage: 0.0%
INFO 07-15 17:07:50 metrics.py:334] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 3548.2 tokens/s, Running: 2041 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 44.2%, CPU KV cache usage: 0.0%
INFO 07-15 17:07:55 metrics.py:334] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 3437.2 tokens/s, Running: 2041 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 44.2%, CPU KV cache usage: 0.0%
Processed prompts: 100%|████████████████████████████████████████████████████████████████████████| 2048/2048 [01:22<00:00, 24.82it/s]
```

</details>

<details>
<summary>optimized version</summary>

```
INFO 07-15 17:20:55 metrics.py:295] Avg prompt throughput: 31913.1 tokens/s, Avg generation throughput: 31.2 tokens/s, Running: 160 reqs, Swapped: 0 reqs, Pending: 1888 reqs, GPU KV cache usage: 3.1%, CPU KV cache usage: 0.0%.
INFO 07-15 17:21:00 metrics.py:295] Avg prompt throughput: 190974.6 tokens/s, Avg generation throughput: 186.5 tokens/s, Running: 1120 reqs, Swapped: 0 reqs, Pending: 928 reqs, GPU KV cache usage: 21.5%, CPU KV cache usage: 0.0%.
INFO 07-15 17:21:05 metrics.py:295] Avg prompt throughput: 187246.9 tokens/s, Avg generation throughput: 182.9 tokens/s, Running: 2048 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 39.3%, CPU KV cache usage: 0.0%.
INFO 07-15 17:21:10 metrics.py:295] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6416.5 tokens/s, Running: 2046 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 39.9%, CPU KV cache usage: 0.0%.
INFO 07-15 17:21:16 metrics.py:295] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6046.0 tokens/s, Running: 2045 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 40.5%, CPU KV cache usage: 0.0%.
INFO 07-15 17:21:21 metrics.py:295] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6079.2 tokens/s, Running: 2045 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 41.1%, CPU KV cache usage: 0.0%.
INFO 07-15 17:21:26 metrics.py:295] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 5687.9 tokens/s, Running: 2045 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 41.7%, CPU KV cache usage: 0.0%.
INFO 07-15 17:21:31 metrics.py:295] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 5591.1 tokens/s, Running: 2045 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 42.3%, CPU KV cache usage: 0.0%.
INFO 07-15 17:21:36 metrics.py:295] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 4830.2 tokens/s, Running: 2044 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 42.9%, CPU KV cache usage: 0.0%.
INFO 07-15 17:21:42 metrics.py:295] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 5256.2 tokens/s, Running: 2043 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 43.5%, CPU KV cache usage: 0.0%.
INFO 07-15 17:21:47 metrics.py:295] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 5286.0 tokens/s, Running: 2042 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 44.1%, CPU KV cache usage: 0.0%.
INFO 07-15 17:21:52 metrics.py:295] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 5070.9 tokens/s, Running: 0 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.0%, CPU KV cache usage: 0.0%.
Processed prompts: 100%|█| 2048/2048 [00:57<00:00, 35.36it/s, est. speed input: 36204.36 toks/s, output: 4517.
```

</details>

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
This seems to share quite a bit in common with `vllm.utils.make_tensor_with_pad` (recently optimized by #4196). I wonder why the utils function wasn't being used in the first place?

File: vllm/model_executor/sampling_metadata.py [hunk 1]
```diff
@@ -517,18 +522,8 @@ def from_lists(cls, temperatures: List[float], top_ps: List[float],
...
-                pin_memory=pin_memory,
-            )
-            output_tensor = torch.tensor(
-                output_padded_tokens,
-                device="cpu",
-                dtype=torch.long,
-                pin_memory=pin_memory,
-            )
+            prompt_tensor = torch.from_numpy(prompt_padded_tokens)
+            output_tensor = torch.from_numpy(output_padded_tokens)
```
Comment (comaniac) [inline vllm/model_executor/sampling_metadata.py:526]: Do we need to add .pin_memory() when available? It should help CPU-GPU data transfer performance, but introduces another data copy to the pin memory.

Comment (Yard1) [reply vllm/model_executor/sampling_metadata.py:526]: we definitely should add that

Comment (peng1999) [reply vllm/model_executor/sampling_metadata.py]: Addressed in 93203016.

**Review by comaniac — APPROVED** [review]

File: vllm/model_executor/sampling_metadata.py [hunk 2]
```diff
@@ -517,18 +522,8 @@ def from_lists(cls, temperatures: List[float], top_ps: List[float],
...
-                pin_memory=pin_memory,
-            )
-            output_tensor = torch.tensor(
-                output_padded_tokens,
-                device="cpu",
-                dtype=torch.long,
-                pin_memory=pin_memory,
-            )
+            prompt_tensor = torch.from_numpy(prompt_padded_tokens).pin_memory()
+            output_tensor = torch.from_numpy(output_padded_tokens).pin_memory()
```
Comment (DarkLight1337) [inline vllm/model_executor/sampling_metadata.py:526]: To keep the old logic, we should only pin memory, if the `pin_memory` variable is `True`

Comment (peng1999) [reply vllm/model_executor/sampling_metadata.py]: Addressed in e7f3c9ddb0d9a96faf336854ef340f9155da72ab.
